src/WinUsb/WinUsb.Helpers.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright © .NET Foundation and Contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace PInvoke
+{
+    using System;
+    using System.Runtime.InteropServices;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <content>
+    /// Methods and nested types that are not strictly P/Invokes but provide
+    /// a slightly higher level of functionality to ease calling into native code.
+    /// </content>
+    public static partial class WinUsb
+    {
+        /// <summary>
+        /// Asynchronously reads data from the specified pipe.
+        /// </summary>
+        /// <param name="interfaceHandle">
+        /// An opaque handle to the interface that contains the endpoint with which the pipe is associated.
+        /// </param>
+        /// <param name="pipeID">
+        /// <paramref name="pipeID"/> corresponds to the bEndpointAddress field in the endpoint descriptor.
+        /// </param>
+        /// <param name="buffer">
+        /// A caller-allocated buffer that receives the data that is read.
+        /// </param>
+        /// <param name="cancellationToken">
+        /// A <see cref="CancellationToken"/> which can be used to cancel the asynchronous operation.
+        /// </param>
+        /// <returns>
+        /// A <see cref="ValueTask"/> which represents the asynchronous operation, and returns the number of bytes transferred.
+        /// </returns>
+        public static unsafe ValueTask<int> WinUsb_ReadPipeAsync(SafeUsbHandle interfaceHandle, byte pipeID, Memory<byte> buffer, CancellationToken cancellationToken)
+        {
+            var overlapped = new WinUsbOverlapped(interfaceHandle, pipeID, buffer);
+            var nativeOverlapped = overlapped.Pack();
+
+            cancellationToken.Register(overlapped.Cancel);
```

Comment:
`Register` returns a struct that should be disposed after the operation to be canceled is over to avoid a memory leak.

Reply:
Thanks, I passed the `CancellationToken` to the `WinUsbOverlapped` class and moved the code calls `CancellationToken.Register` / `CancellationTokenRegistration.Dispose` to `WinUsbOverlapped.Pack` and `WinUsbOverlapped.Unpack`